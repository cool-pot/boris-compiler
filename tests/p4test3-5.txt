*** test level 3,  local foreach range
defun arrayfunc(x)
    foreach k in 5..20 do
        print k*k;
    end for
    return 0;
end defun

print arrayfunc 0;