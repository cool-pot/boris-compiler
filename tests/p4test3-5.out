; ModuleID = 'module-boris-input'
source_filename = "module-boris-input"
target triple = "x86_64-pc-linux-gnu"

@format_int = private unnamed_addr constant [4 x i8] c"%d\0A\00"
@format_int.1 = private unnamed_addr constant [4 x i8] c"%d\0A\00"

define i32 @strlen(i8* %s) {
init:
  %i = alloca i32
  store i32 0, i32* %i
  br label %check

check:                                            ; preds = %body, %init
  %0 = load i32, i32* %i
  %1 = getelementptr i8, i8* %s, i32 %0
  %2 = load i8, i8* %1
  %3 = icmp ne i8 0, %2
  br i1 %3, label %body, label %end

body:                                             ; preds = %check
  %4 = load i32, i32* %i
  %5 = add i32 %4, 1
  store i32 %5, i32* %i
  br label %check

end:                                              ; preds = %check
  %6 = load i32, i32* %i
  ret i32 %6
}

declare i32 @printf(...)

define void @main() {
MainEntry:
  %0 = call i32 @arrayfunc(i32 0)
  %1 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int.1, i32 0, i32 0), i32 %0)
  br label %EndEntry

EndEntry:                                         ; preds = %MainEntry
  ret void
}

define i32 @arrayfunc(i32) {
Entry:
  %x = alloca i32
  store i32 %0, i32* %x
  br label %foreachr_init_

foreachr_init_:                                   ; preds = %Entry
  %k = alloca i32
  store i32 5, i32* %k
  br label %foreachr_check_

foreachr_check_:                                  ; preds = %foreachr_body_, %foreachr_init_
  %1 = load i32, i32* %k
  %2 = icmp sle i32 %1, 20
  br i1 %2, label %foreachr_body_, label %foreachr_end_

foreachr_body_:                                   ; preds = %foreachr_check_
  %load_singleid_int_temp = load i32, i32* %k
  %load_singleid_int_temp1 = load i32, i32* %k
  %expr_mul_expr_tmp = mul i32 %load_singleid_int_temp, %load_singleid_int_temp1
  %3 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int, i32 0, i32 0), i32 %expr_mul_expr_tmp)
  %4 = load i32, i32* %k
  %5 = add i32 %4, 1
  store i32 %5, i32* %k
  br label %foreachr_check_

foreachr_end_:                                    ; preds = %foreachr_check_
  ret i32 0
}
